src/freenet/clients/http/ToadletContextImpl.java

```diff
@@ -376,15 +376,32 @@ static void sendReplyHeaders(OutputStream sockOutputStream, int replyCode, Strin
 			}
 		if(contentLength >= 0)
 			mvt.put("content-length", Long.toString(contentLength));
-		
-		String expiresTime;
+
+		boolean allowCaching; // For privacy reasons, only static
+							  // content may be cached
 		if (mTime == null) {
-			expiresTime = "Thu, 01 Jan 1970 00:00:00 GMT";
+			allowCaching = false;
+		} else {
+			allowCaching = true;
+		}
+		String expiresTime;
+		String cacheControl;
+		if (allowCaching) {
+			// use an expiry time of 30 day from now, about the frequency of Freenet releases
+			// Expires is needed for older browsers
+			expiresTime = TimeUtil.makeHTTPDate(System.currentTimeMillis() + DAYS.toMillis(30));
+			cacheControl = "public, max-age=" + String.valueOf(3600 * 24 * 30);
+			// Use contentlength as simplest possible hash function
+			// which needs no additional file access
+			mvt.put("ETag", String.valueOf(contentLength));
```

Comment:
@ArneBab why would you want to use something that is not guaranteed to change when a file changes when you have something available that is guaranteed to change when the file changes? That’s what the ETag is about: detecting changes. If you are really worried about caching files across restarts, use the hash. Stop trying to find supposedly cheaper shortcuts that will only make things worse by royally messing up the caching!